### ISSUE TITLE
transaction_index::index and batched calls

### ISSUE THREAD

**Issue by sinkingsugar** [paritytech/substrate#12835]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

Any extrinsic using `transaction_index::index` will not work properly if used within a `pallet_utility` batch call as it depends on extrinsic index and during a batch there is only one index.
In our chain this results into corrupted assets without any way to know about it.

Overall `transaction_index::index` is quite rudimentary as it relies on a inverted offset of bytes from call last byte...
This is an amazing feature allowing on chain storage accessed via IPFS which would totally use more love.

Extra Question:
Is there a way to detect if the extrinsic is being batched? This would be enough to actually make it fail...

### Steps to reproduce

Just batch a bunch of `frame/transaction-storage` extrinsics

**Comment by sinkingsugar** [comment]
> > Is there a way to detect if the extrinsic is being batched? This would be enough to actually make it fail...
> 
> 
> 
> If you don't want to support batching, you could filter them in your `BaseCallFilter`.

Well they are a good feature. Maybe we can filter them out for certain calls though! That's probably an idea worth investigating, thanks 🙏!

**Comment by bkchr** [comment]
Did it work?

**Comment by sinkingsugar** [comment]
> Did it work?

Not sure, @aang114 working on it, I think he is asking you many questions already on the stack overflow thing :)
But I think it will work, I don't see why it should not, we are actually going to block also too deep call stacks as our transactions can be very big ( we break the regular limits on our chain )

**Comment by bkchr** [comment]
Okay, please close the issue if you think that is solved or get back with further questions :)

**Comment by sinkingsugar** [comment]
I think the issue is still there, what you proposed is a good workaround to avoid polluting the chain with invalid data, but won't change the fact `transaction_index::index` can't be used when the call is nested inside a batch. So I'd say to keep it open at least as way to let new users know that there is this issue and it's not resolved (I understand priority is low and work wise it's quite some refactor).

**Comment by bkchr** [comment]
Maybe @arkpar can give some feedback here.

**Comment by aang114** [comment]
> Did it work?

Yes, we are using `BaseCallFilter` to prevent a **batch call (i.e `pallet_utility::batch()`) that includes a call that uses `transaction_index::index`** from **ever being dispatched**.

However, we would also like to prevent any extrinsic that contains "a batch call (i.e `pallet_utility::batch()`) that includes a call that uses `transaction_index::index`" from **being gossiped to other nodes**. But right now the only solution I see is using `<Runtime as TaggedTransactionQueue<Block>>::validate_transaction()` and writing recursive logic (recursive since we need to check all the inner calls) to ensure this. Would you recommend this approach? Thanks

**Comment by bkchr** [comment]
Why do you want to prevent that this is gossiped? If someone sends this transaction, they will need to pay fees for the transaction and it will return an error. This should de-incentivize people to do this and if they still do it they need to pay and don't get anything.

**Comment by aang114** [comment]
> Why do you want to prevent that this is gossiped? If someone sends this transaction, they will need to pay fees for the transaction and it will return an error. This should de-incentivize people to do this and if they still do it they need to pay and don't get anything.

True. But initially, we will be waiving off transaction fees for certain users. Therefore, we want to prevent incorrect transactions from being gossiped. I guess `<Runtime as TaggedTransactionQueue<Block>>::validate_transaction()` is the best way to about this then, right?

**Comment by aang114** [comment]
> > Why do you want to prevent that this is gossiped? If someone sends this transaction, they will need to pay fees for the transaction and it will return an error. This should de-incentivize people to do this and if they still do it they need to pay and don't get anything.
> 
> True. But initially, we will be waiving off transaction fees for certain users. And furthermore, a transaction's length on our blockchain can be really huge (it can potentially fill up the entire block). Therefore, we don't want our blocks to be polluted/spammed with invalid transactions - since this can essentially "stall our blockchain" (since genuine/valid transactions might take a lot more time to get processed).
> 
> And therefore, we want to even prevent incorrect transactions from being gossiped. I guess writing a recursive check (recursive since we need to check all the inner calls) in `<Runtime as TaggedTransactionQueue<Block>>::validate_transaction()` is the best way to go about this then, right?

@bkchr Btw this is our (recursive) solution for ensuring that an extrinsic does not contain an invalid call or does not exceed maximum depth length **before it is gossiped**: https://github.com/fragcolor-xyz/clamor/blob/865a106592bd5ce4a8b4c4b64501e82dfc21eb2a/runtime/src/lib.rs#L1297-L1308.

Please let us know if we could optimize further. Thanks again! 🙂
